apostrophecms/apostrophe
# Advanced build optimization

## Summary

Add:
Improve the build task to accept a changes parameter and based on that to detect what build to execute. Improve the build watcher to send the paths of the detected changes and restart the application only when needed. Allow external chokidar watcher to be registered. The smart building of Apostrophe UI happens on relevant change, no matter if APOS_DEV is enabled.

Fix:
- Fix a bug where some Apostrophe modules symlinked in node_modules are not being watched.
- Recover after a build error when in a watch mode. 

## What are the specific steps to test this change?

1. Start Apostrophe application in dev mode: `node app`
- the old build behavior is expected (execute `src` related and `public` builds, complie `apos` based on `package-lock.json`)
- a "watching files" message should be visible in the console
- a change to `ui/src` should re-concatenate all bundles but only recompile the `ui/src` portion
- a change to `ui/public` should re-concatenate everything but no Webpack builds should be executed
- a change to `ui/apos` should only rebuild the apos bundle
- edit any watched asset and create a bad code to reproduce Webpack build error
- console log should show detailed information about the error
- the page should not be reloaded
- edit the same file and correct the error
- a normal rebuild should occur
- the page should be reloaded 
2. Start Apostrophe application in dev mode: `APOS_DEV=1 node app`
- the old build behavior is expected (rebuild everything)
- same as 1.
3. Start Apostrophe application in dev mode: `APOS_ASSET_WATCH=0 node app`
- a "watching files" message should NOT be visible in the console
- the old build behavior is expected (execute `src` related and `public` builds, complie `apos` based on `package-lock.json`)
- after the initial build, changes should not trigger any build

## What kind of change does this PR introduce?
*(Check at least one)*


## Make sure the PR fulfills these requirements:
